File: lintcode/reverse-linked-list.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/reverse-linked-list/

from linked_list import *

"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: You should return the head of the reversed linked list.
                  Reverse it in-place.
    """
    def reverse(self, head):
        # write your code here
        dummy = ListNode(0)
        reader = head
        while reader:
            temp = reader.next
            reader.next = dummy.next
            dummy.next = reader
            reader = temp
        return dummy.next

if __name__ == '__main__':
    s = Solution()
    head = create_linked_list([1,2,3])
    new_head = s.reverse(head)
    print_linked_list(new_head)

File: leetcode-2/reverse-linked-list.cc
// Question: https://leetcode.com/problems/reverse-linked-list/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode dummy(0);
        auto cursor = &dummy;
        auto node = head;
        while (node) {
            auto next = node->next;
            node->next = cursor->next;
            cursor->next = node;
            node = next;
        }
        return dummy.next;
    }

    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }

        auto ans = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        return ans;
    }
};
